### PR TITLE
Add warning message when trying to navigate away when creating/editing an action

### LINKED ISSUE #30139: FE - Creating or editing an action

### PR DESCRIPTION
**Pull request by oisincoveney**
### Description

Adds a `beforeunload` event when the user has created or is editing an action, so the user will be prompted with an "Are you sure" browser prompt when they try to refresh/close the tab/navigate to a new URL.

#### Basic / Implicit Actions
1. Create or access a model that has actions enabled.
2. If you haven't already, enable basic actions (go to menu icon in top right corner of the actions container and click "enable basic actions"
3. Try to edit the success message by editing a basic action, clicking "Action Settings" and changing the message
4. Try to refresh/close/navigate away from the page without hitting "Update" 

#### Custom Actions
_Creating an action_
1. Create an action - I did this by accessing a model, then going to the `Model Details > Actions`, then creating an action there.
2. Try to refresh/close/navigate away from the page. 

_Editing an action_

1. Go to an existing action (maybe the one that you created above) and try to edit the action.
2. Make a small change to either the query text or its parameter metadata.
5. Try to refresh/close/navigate away from the page. 

_In all cases, you should see something similar to the screenshot below._

### Demo

Screenshot 2023-04-21 at 5 04 36 PM

### Checklist

- [x] Tests have been added/updated to cover changes in this PR

- - -
This change is [Reviewable](https://reviewable.io/reviews/metabase/metabase/30292)

### CODE REVIEW

File: frontend/src/metabase/actions/containers/ActionCreator/FormCreator/FormCreator.styled.tsx [hunk 1]
```diff
@@ -40,7 +40,10 @@ export const ExplainerTitle = styled.h3`
   margin-bottom: ${space(1)};
 `;
 
-export const ExplainerText = styled.p`
+// REMINDER: mention that using `styled.p` causes a
```
Comment (oisincoveney) [inline frontend/src/metabase/actions/containers/ActionCreator/FormCreator/FormCreator.styled.tsx:43]: What do y'all think? I think we should try to keep our HTML semantically useful but having the `p > ul` is causing errors, so I'm not sure we should keep this as a `p`

**Review by WiNloSt — COMMENTED** [review]
> Reviewed 16 of 16 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/metabase/metabase/30292)* status: all files reviewed, 4 unresolved discussions (waiting on @kulyk and @oisincoveney)

___
*[`frontend/src/metabase/actions/containers/ActionCreator/ActionContext/QueryActionContextProvider/QueryActionContextProvider.tsx` line 180 at r3](https://reviewable.io/reviews/metabase/metabase/30292#-NTsaO_s4GNWSIEpTIrf:-NTsaO_s4GNWSIEpTIrg:b-un2z17) ([raw file](https://github.com/metabase/metabase/blob/79d2263795181da44c34ade1e82f82b01477e60d/frontend/src/metabase/actions/containers/ActionCreator/ActionContext/QueryActionContextProvider/QueryActionContextProvider.tsx#L180)):*
> ```XML
>   );
> 
>   const isDirty = useMemo(() => {
> ```

Should we use similar logic from `ImplicitActionContextProvider`? That file passes `canSave` to `isDirty`. These 2 actions behave a bit differently.

1. Implicit actions don't allow us to update them when there's no change
    
2. Query actions still allow us to update them even if nothing has changed.
    

You can see in the examples above that implicit actions disable the update button when nothing changed, but not the query actions.

If we make this behavior the same we could use `canSave` as `isDirty`. @kulyk do you have any insight why these behaviors are different for implicit vs query actions?
___
*[`frontend/src/metabase/actions/containers/ActionCreator/ActionContext/QueryActionContextProvider/QueryActionContextProvider.tsx` line 199 at r3](https://reviewable.io/reviews/metabase/metabase/30292#-NTsciHR5q4_Vg8nJMmq:-NTsciHR5q4_Vg8nJMmr:b1gecuq) ([raw file](https://github.com/metabase/metabase/blob/b6b9eb3763c0aa422dcb8ad3fe256f2b70a55c3a/frontend/src/metabase/actions/containers/ActionCreator/ActionContext/QueryActionContextProvider/QueryActionContextProvider.tsx#L199)):*
> ```XML
>       canSave,
>       handleActionChange,
>       setFormSettings,
> ```

:thumbsup: Not sure why this is here in the first place, since it's a state setter from `useState` :thinking: 
___
*[`frontend/src/metabase/actions/containers/ActionCreator/FormCreator/FormCreator.styled.tsx` line 43 at r3](https://reviewable.io/reviews/metabase/metabase/30292#gh-1176276196:-NTsdHq0Bkh-5s3ZWdq8:bwmeqrh) ([raw file](https://github.com/metabase/metabase/blob/79d2263795181da44c34ade1e82f82b01477e60d/frontend/src/metabase/actions/containers/ActionCreator/FormCreator/FormCreator.styled.tsx#L43)):*
<details><summary><i>Previously, oisincoveney (Oisin Coveney) wrote…</i></summary><blockquote>

What do y'all think? I think we should try to keep our HTML semantically useful but having the `p > ul` is causing errors, so I'm not sure we should keep this as a `p`
</blockquote></details>

Sounds good for keeping it semantically correct. But I'm not sure that reason warrants this comment's existence in the code. What do you think?
___
*[`frontend/src/metabase/actions/containers/ActionCreator/FormCreator/FormCreator.styled.tsx` line 70 at r3](https://reviewable.io/reviews/metabase/metabase/30292#-NTseGNc3joBlwD0vLNy:-NTseGNc3joBlwD0vLNz:b-xb79iy) ([raw file](https://github.com/metabase/metabase/blob/79d2263795181da44c34ade1e82f82b01477e60d/frontend/src/metabase/actions/containers/ActionCreator/FormCreator/FormCreator.styled.tsx#L70)):*
> ```XML
> 
>   color: ${color("brand")};
> ```

nit: if this is not formatted automatically I'd leave this change out just to keep the git history saner.

**Review by oisincoveney — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/metabase/metabase/30292)* status: 15 of 16 files reviewed, 4 unresolved discussions (waiting on @kulyk, @oisincoveney, and @WiNloSt)

___
*[`frontend/src/metabase/actions/containers/ActionCreator/ActionContext/QueryActionContextProvider/QueryActionContextProvider.tsx` line 180 at r3](https://reviewable.io/reviews/metabase/metabase/30292#-NTsaO_s4GNWSIEpTIrf:-NTsnN6w3-ILxk1aOLXd:b-w3pclg) ([raw file](https://github.com/metabase/metabase/blob/79d2263795181da44c34ade1e82f82b01477e60d/frontend/src/metabase/actions/containers/ActionCreator/ActionContext/QueryActionContextProvider/QueryActionContextProvider.tsx#L180)):*
<details><summary><i>Previously, WiNloSt (Mahatthana (Kelvin) Nomsawadi) wrote…</i></summary><blockquote>

Should we use similar logic from `ImplicitActionContextProvider`? That file passes `canSave` to `isDirty`. These 2 actions behave a bit differently.

1. Implicit actions don't allow us to update them when there's no change
    
2. Query actions still allow us to update them even if nothing has changed.
    

You can see in the examples above that implicit actions disable the update button when nothing changed, but not the query actions.

If we make this behavior the same we could use `canSave` as `isDirty`. @kulyk do you have any insight why these behaviors are different for implicit vs query actions?
</blockquote></details>

Gotcha - just for context I noticed that the behavior for QueryActionContextProvider is a lot more involved than the Implicit actions - but I also didn't want to change `canSave` since it's just a check on whether the query is empty. I felt adding the extra logic to the Query ContextProvider made sense, and using `canSave` in the Implicit ContextProvider would do the trick since there's a lot less available there. But would love to hear Anton's input on this as well, I found it a bit jarring when I was working with the different actions for the first time
___
*[`frontend/src/metabase/actions/containers/ActionCreator/ActionContext/QueryActionContextProvider/QueryActionContextProvider.tsx` line 199 at r3](https://reviewable.io/reviews/metabase/metabase/30292#-NTsciHR5q4_Vg8nJMmq:-NTsn8d4CmVKos65A-QC:b-xw5uve) ([raw file](https://github.com/metabase/metabase/blob/b6b9eb3763c0aa422dcb8ad3fe256f2b70a55c3a/frontend/src/metabase/actions/containers/ActionCreator/ActionContext/QueryActionContextProvider/QueryActionContextProvider.tsx#L199)):*
<details><summary><i>Previously, WiNloSt (Mahatthana (Kelvin) Nomsawadi) wrote…</i></summary><blockquote>

:thumbsup: Not sure why this is here in the first place, since it's a state setter from `useState` :thinking: 
</blockquote></details>

Yep I agree, I really don't know how that got there
___
*[`frontend/src/metabase/actions/containers/ActionCreator/FormCreator/FormCreator.styled.tsx` line 43 at r3](https://reviewable.io/reviews/metabase/metabase/30292#gh-1176276196:-NTsoPOx24-Y-glfQxMZ:buolrlp) ([raw file](https://github.com/metabase/metabase/blob/79d2263795181da44c34ade1e82f82b01477e60d/frontend/src/metabase/actions/containers/ActionCreator/FormCreator/FormCreator.styled.tsx#L43)):*
<details><summary><i>Previously, WiNloSt (Mahatthana (Kelvin) Nomsawadi) wrote…</i></summary><blockquote>

Sounds good for keeping it semantically correct. But I'm not sure that reason warrants this comment's existence in the code. What do you think?
</blockquote></details>

Let's definitely take it out - I just knew that if I didn't remind myself with the comments while I was coding, I would forget to ask 😅

**Review by WiNloSt — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/metabase/metabase/30292)* status: 15 of 16 files reviewed, 4 unresolved discussions (waiting on @kulyk and @oisincoveney)

___
*[`frontend/src/metabase/actions/containers/ActionCreator/FormCreator/FormCreator.styled.tsx` line 43 at r3](https://reviewable.io/reviews/metabase/metabase/30292#gh-1176276196:-NTss35q21nu1JQUZ3Bu:besreyb) ([raw file](https://github.com/metabase/metabase/blob/79d2263795181da44c34ade1e82f82b01477e60d/frontend/src/metabase/actions/containers/ActionCreator/FormCreator/FormCreator.styled.tsx#L43)):*
<details><summary><i>Previously, oisincoveney (Oisin Coveney) wrote…</i></summary><blockquote>

Let's definitely take it out - I just knew that if I didn't remind myself with the comments while I was coding, I would forget to ask 😅
</blockquote></details>

I do that too! I leave comments with a prefixe that I could look up later and remove once I've added comments in my PRs. e.g. [this commit](https://github.com/metabase/metabase/pull/29569/commits/3ceea502393779b072a731169a14744b47924315)

actually I use this [VS Code extension](https://marketplace.visualstudio.com/items?itemName=Gruntfuggly.todo-tree) to list all the comment I have left in the codebase, so I can never forget to remove them :)

**Comment by uladzimirdev** [comment]
*[`frontend/src/metabase/actions/containers/ActionCreator/FormCreator/FormCreator.styled.tsx` line 43 at r3](https://reviewable.io/reviews/metabase/metabase/30292#gh-1176276196:-NTwoRJxAHGWXfzJNhJs:b-g6l8yk) ([raw file](https://github.com/metabase/metabase/blob/79d2263795181da44c34ade1e82f82b01477e60d/frontend/src/metabase/actions/containers/ActionCreator/FormCreator/FormCreator.styled.tsx#L43)):*
<details><summary><i>Previously, WiNloSt (Mahatthana (Kelvin) Nomsawadi) wrote…</i></summary><blockquote>

I do that too! I leave comments with a prefixe that I could look up later and remove once I've added comments in my PRs. e.g. https://github.com/metabase/metabase/pull/29569/commits/3ceea502393779b072a731169a14744b47924315

actually I use this [VS Code extension](https://marketplace.visualstudio.com/items?itemName=Gruntfuggly.todo-tree) to list all the comment I have left in the codebase, so I can never forget to remove them :)
</blockquote></details>

@kulyk recommended me to use github comments instead, but personally I have the same problem - usually I forget to ask if I do not leave a comment in the code 

nice extension @WiNloSt

**Review by uladzimirdev — COMMENTED** [review]
> Reviewed 15 of 16 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/metabase/metabase/30292)* status: all files reviewed, 4 unresolved discussions (waiting on @kulyk, @oisincoveney, and @WiNloSt)

**Review by oisincoveney — COMMENTED** [review]
> Dismissed @WiNloSt from a discussion.
> *[Reviewable](https://reviewable.io/reviews/metabase/metabase/30292)* status: all files reviewed, 3 unresolved discussions (waiting on @kulyk and @oisincoveney)

___
*[`frontend/src/metabase/actions/containers/ActionCreator/FormCreator/FormCreator.styled.tsx` line 43 at r3](https://reviewable.io/reviews/metabase/metabase/30292#gh-1176276196:-NTxMjiY0d5Lm1shpEd8:b46p0n6) ([raw file](https://github.com/metabase/metabase/blob/79d2263795181da44c34ade1e82f82b01477e60d/frontend/src/metabase/actions/containers/ActionCreator/FormCreator/FormCreator.styled.tsx#L43)):*
<details><summary><i>Previously, uladzimirdev (Uladzimir Havenchyk) wrote…</i></summary><blockquote>

@kulyk recommended me to use github comments instead, but personally I have the same problem - usually I forget to ask if I do not leave a comment in the code 

nice extension @WiNloSt 
</blockquote></details>

Awesome! I'm usually an IntellJ type of guy (I know, I know :smile: ), I'll see if there's anything similar

**Review by WiNloSt — APPROVED** [review]
Great work! Tested and this works great.

> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/metabase/metabase/30292#-:-NU1drgC52ufYU-hOx9D:b-wp31db)* status: all files reviewed, 1 unresolved discussion (waiting on @kulyk)

___
*[`frontend/src/metabase/actions/containers/ActionCreator/ActionContext/QueryActionContextProvider/QueryActionContextProvider.tsx` line 180 at r3](https://reviewable.io/reviews/metabase/metabase/30292#-NTsaO_s4GNWSIEpTIrf:-NU1deoYDXTe22xEPZ5N:brvhqxx) ([raw file](https://github.com/metabase/metabase/blob/79d2263795181da44c34ade1e82f82b01477e60d/frontend/src/metabase/actions/containers/ActionCreator/ActionContext/QueryActionContextProvider/QueryActionContextProvider.tsx#L180)):*
<details><summary><i>Previously, oisincoveney (Oisin Coveney) wrote…</i></summary><blockquote>

Gotcha - just for context I noticed that the behavior for QueryActionContextProvider is a lot more involved than the Implicit actions - but I also didn't want to change `canSave` since it's just a check on whether the query is empty. I felt adding the extra logic to the Query ContextProvider made sense, and using `canSave` in the Implicit ContextProvider would do the trick since there's a lot less available there. But would love to hear Anton's input on this as well, I found it a bit jarring when I was working with the different actions for the first time
</blockquote></details>

I don't want this discussion to block this PR. I'll approve this, but it'd be great to hear from @kulyk and possibly follow this up in another PR if needed.

**Pull request merged**